### ISSUE TITLE
`fan` for `HashMap`

### ISSUE THREAD

**Issue by dalaing** [reflex-frp/reflex#311]
Someone was asking about this on IRC, and it got me to wondering:  would it be useful to add a primitive like `fan` but using something like `HashMap` instead of `Map`?

This would probably require writing something like `dependent-map` but for `unordered-containers`, and new primitives in the spider, along with the userland API to make it all go.

It sounds like it might be a lot of work, so I wonder how much of a performance win would be needed to justify it, and whether there are small prototype-y things that could be done to gather a bit more confidence about the amount of work and/or effect on performance.

**Comment by ryantrinkle** [comment]
It'd definitely be useful, but the real difficulty here is in abstracting the underlying primitives.  Really, they shouldn't care about the details of the container - they should just demand some typeclass, with things like union and intersection.  However, I haven't determined what exact laws and operations they would require.

All of this also applies to `merge`.

**Comment by ryantrinkle** [comment]
It would be great to have someone's help on this.  As far as Spider core issues are concerned, it sounds relatively approachable, and I'd be happy to help someone dig into it any time after this month.

**Comment by adamConnerSax** [comment]
I don't think this solves this at as deep a level as you're looking for and it might be too complicated, but it does, perhaps, address the original issue. 
https://github.com/adamConnerSax/reflex-collections/blob/master/src/Reflex/Collections/ToPatchType.hs

**Comment by dalaing** [comment]
I'm actually keen to take a run at this work for `DHashMap` and working out how to implement the equivalents of `fan` and `merge` in the Spider.  If it turns out to be a big performance win, that would be great, but I think I can justify the time as a useful exercise for myself even with out that - which is nice, because it means that I wouldn't be sweating about whether the performance was going to justify the work as I was doing it :)

**Comment by adamConnerSax** [comment]
Cool!  From the bit I looked at--not at the Spider level, but above it where intMap and Map were supported--I agree with Ryan that the essence of at least part of it is figuring out how to generalize the idea of a "diff" and the application of that diff as a patch.  
Also, though this may have been an artifact of the level at which I was approaching things, I struggled a bunch with the fact that some of the algos required an empty container.  That made it difficult to include things like Arrays and some sorts of Trees, though those are clearly useful.  I worked around it with Sum types to add an empty case but that led to other messiness...
Anyway, I'm curious to see what you come up with!

**Comment by treeowl** [comment]
Even a non-dependent `HashMap` would be quite useful to have. I'm running into that right now....

**Comment by adamConnerSax** [comment]
Take a look at what I linked above. It might suit.
